Set/remove the Content-Type when mutating the method

Leyart opened the pull request:
Opening a GET request with the request editor and switching the method with the dropdown from GET to POST moves the query parameters to the request body but it is not adding the correct content-type header, making the call silently fail.
This PR sets correctly the content-type header when switching methods in the requester add-on.

Fix zaproxy/zaproxy#7440

Signed-off-by: Leyart [leyart@gmail.com](mailto:leyart@gmail.com)

thc202 commented on zap/src/main/java/org/parosproxy/paros/network/HttpMessage.java:
```diff
@@ -990,6 +990,11 @@ public void mutateHttpMethod(String method) {
             if (prevMethod.equalsIgnoreCase(method)) {
                 return;
             }
+            // When changing methods, remove the content-type if a method should not have a body
+            // This does not prevent the user for modifying the request afterwards if needed
```
Don't add this.

thc202 commented on zap/src/main/java/org/parosproxy/paros/network/HttpMessage.java:
```diff
@@ -990,6 +990,11 @@ public void mutateHttpMethod(String method) {
             }
+            // When changing methods, remove the content-type if a method should not have a body
+            // This does not prevent the user for modifying the request afterwards if needed
+            if (!(method.equals(HttpRequestHeader.POST) || method.equals(HttpRequestHeader.PUT))) {
```
IMO we should only change the Content-Type if we are actually changing the parameters around. (This also misses methods where no body is expected.)

Leyart replied:
I am not sure if I understood the comment here. 
I have put this check because if we are switching from a POST or PUT back to a GET (or any other method that does not have a body), there should be no Content-Type set, since all the parameters are moved back as query parameters. That's why by setting it to null we are removing it.

thc202 replied:
Mutating from PUT to GET does not change anything. The body is not changed unless mutating from/to POST, so we shouldn't be changing unless it's those cases.

Leyart replied:
Thanks for the clarification, it makes sense. I will just remove the content-type if the previous method is POST then

thc202 commented on zap/src/main/java/org/parosproxy/paros/network/HttpMessage.java:
```diff
@@ -1035,6 +1040,11 @@ public void mutateHttpMethod(String method) {
                             sb.append('=');
                         }
                     }
+                    // Explicitly set the Content-Type for the body of the POST
```
Don't add this.

thc202 commented on zap/src/test/java/org/parosproxy/paros/network/HttpMessageUnitTest.java:
```diff
@@ -273,6 +273,45 @@ void shouldNotSetContentEncodingsWhenSettingHttpRequestBody() {
     }
 
+    @Test
+    void shouldSetContentTypeWhenSettingMethodToPOST() throws Exception {
```
- `Setting` → `Mutating`
- `ToPOST` → `ToPost`

thc202 commented on zap/src/test/java/org/parosproxy/paros/network/HttpMessageUnitTest.java:
```diff
@@ -273,6 +273,45 @@ void shouldNotSetContentEncodingsWhenSettingHttpRequestBody() {
+    }
+
+    @Test
+    void shouldRemoveContentTypeWhenSettingMethodToGET() throws Exception {
```
- `Setting` → `Mutating`
- `ToGET` → `FromPost`
- Should be parameterized.

thc202 replied:
The parameterization should be on the method.

Leyart commented:
Sorry for addressing the comments so late, I was on holidays. The content-type is not removed just if the previous method was a POST and set accordingly if we mutate to a POST.
The tests are now method parametrized, excluding POST that is set explicitly or mutated from explicitly in the tests.

Leyart commented:
@psiinon @thc202 should be in a good state now, let me know if it needs further adjustments

Leyart commented:
@kingthorin @thc202 @psiinon Happy new year! Just a ping for this PR that should be ready, would be great to be able to use it in the next release.

thc202 commented:
Thank you!
